### PR TITLE
Add config to toggle StrictMode

### LINKED ISSUE #6694: Disable strict mode with a config flag
We add `StrictMode` here for development:

https://github.com/mozilla/addons-frontend/blob/3d8200443ca1734cfbd5d3666a9154d632685881/src/core/components/Root/index.js#L23

This is nice because it warns us when we use deprecated lifecycle methods. However, it also double-triggers lifecycle methods to make sure they are idempotent. This is confusing when you are trying to debug a real problem like making too many API requests -- a common problem if your dispatching logic is incorrect.

I would like a config flag to disable `StrictMode` temporarily in my local environment. After I'm done debugging, I would put it back.

My current workaround is just to edit the code:

```diff
diff --git a/src/core/components/Root/index.js b/src/core/components/Root/index.js
index 634386f04..658985b0e 100644
--- a/src/core/components/Root/index.js
+++ b/src/core/components/Root/index.js
@@ -20,7 +20,7 @@ const Root = ({ children, history, i18n, store }: Props) => (
     <Provider store={store} key="provider">
       <ConnectedRouter history={history}>
         {/* $FLOW_FIXME: https://github.com/facebook/react/issues/12553 */}
-        <React.StrictMode>{children}</React.StrictMode>
+        {children}
       </ConnectedRouter>
     </Provider>
   </I18nProvider>

```

Just as an example, here's how you could use the config to conditionally render `<StrictMode>`:

https://github.com/mozilla/addons-frontend/blob/3d8200443ca1734cfbd5d3666a9154d632685881/src/amo/components/RatingManager/index.js#L343-L345

### PR DESCRIPTION
**Pull request by SeanPrashad**
Fixes #6694: Disable strict mode with a config flag

This PR introduces the ability to toggle [React Strict mode](https://reactjs.org/docs/strict-mode.html), a tool "_for highlighting potential problems in an application_".

### CODE REVIEW

File: src/core/components/Root/index.js [hunk 1]
```diff
@@ -12,18 +13,30 @@ import type { ReactRouterHistoryType } from 'core/types/router';
...
+  store,
+  cookies = null,
+}: Props) => (
   <I18nProvider i18n={i18n}>
     <Provider store={store} key="provider">
       <ConnectedRouter history={history}>
         <CookiesProvider cookies={cookies}>
           {/* $FLOW_FIXME: https://github.com/facebook/react/issues/12553 */}
-          <React.StrictMode>{children}</React.StrictMode>
+          {_config.get('enableStrictMode') ? (
```
Comment (SeanPrashad) [inline src/core/components/Root/index.js]: **Note:** Check out where to set the value `enableStrictMode` as per other examples in one of these files:

image

Comment (SeanPrashad) [reply src/core/components/Root/index.js]: https://travis-ci.org/mozilla/addons-frontend/jobs/458615128#L5310 says that `config` is undefined

**Comment by SeanPrashad** [comment]
Hmm.. there are 2 Assertion and Index errors for the python tests: https://travis-ci.org/mozilla/addons-frontend/jobs/458638843

**Comment by willdurand** [comment]
@SeanPrashad when you declare a new configuration parameter, you have to "white-list" it, see the `clientConfigKeys` key. I think you need to add it here and also in the disco settings. Grep for `enableDevTools` to see where it is registered.

**Comment by SeanPrashad** [comment]
I've added it to the whitelist section in `config/default.js` and `config/default-disco.js` but the errors are still appearing in Travis 🤔

Is there anyway for me to run those tests locally?

File: config/dev.js [hunk 2]
```diff
@@ -9,6 +9,8 @@ module.exports = {
 
   enableDevTools: true,
 
+    enableStrictMode: true,
```
Comment (willdurand) [inline config/dev.js]: Not sure we want it in -dev, especially because it is already `true` by default

File: config/development.js [hunk 3]
```diff
@@ -19,6 +19,8 @@ module.exports = {
 
   enableDevTools: true,
 
+  enableStrictMode: true,
```
Comment (willdurand) [inline config/development.js]: Same here.

File: src/core/components/Root/index.js [hunk 4]
```diff
@@ -9,6 +9,8 @@ import type { I18nType } from 'core/types/i18n';
 import type { ReduxStore } from 'core/types/redux';
 import type { ReactRouterHistoryType } from 'core/types/router';
 
+const config = require('config');
```
Comment (willdurand) [inline src/core/components/Root/index.js]: Can you please use this syntax:

```js
import config from 'config';
```

That's probably what's causing the issue.

Also, I think you can "inject" `config` into the component:

```js
const Root = ({ children, history, i18n, store, cookies = null, _config = config }: Props) => (
```

You'll have to update `Props` with `_config: typeof config`, and use `_config` in L28.

File: src/core/components/Root/index.js [hunk 5]
```diff
@@ -23,7 +25,11 @@ const Root = ({ children, history, i18n, store, cookies = null }: Props) => (
       <ConnectedRouter history={history}>
         <CookiesProvider cookies={cookies}>
           {/* $FLOW_FIXME: https://github.com/facebook/react/issues/12553 */}
-          <React.StrictMode>{children}</React.StrictMode>
+          {config.get('enableStrictMode') ? (
+            <React.StrictMode>{children}</React.StrictMode>
+          ) : (
+            { children }
```
Comment (willdurand) [inline src/core/components/Root/index.js]: Can you remove the spaces around `children`?

Comment (SeanPrashad) [reply src/core/components/Root/index.js]: I tried but ESLint is complaining:

screen shot 2018-11-23 at 1 38 06 pm

File: src/core/components/Root/index.js [hunk 6]
```diff
@@ -3,27 +3,34 @@ import * as React from 'react';
...
 import { CookiesProvider, Cookies } from 'react-cookie';
 import { Provider } from 'react-redux';
+import config from 'config';
 
 import I18nProvider from 'core/i18n/Provider';
 import type { I18nType } from 'core/types/i18n';
 import type { ReduxStore } from 'core/types/redux';
 import type { ReactRouterHistoryType } from 'core/types/router';
 
+
```
Comment (willdurand) [inline src/core/components/Root/index.js]: This should be reverted

File: src/core/components/Root/index.js [hunk 7]
```diff
@@ -3,27 +3,34 @@ import * as React from 'react';
...
 type Props = {|
   children: React.Node,
   cookies?: typeof Cookies,
   history: ReactRouterHistoryType,
   i18n: I18nType,
   store: ReduxStore,
+  _config: typeof config
```
Comment (willdurand) [inline src/core/components/Root/index.js]: Could you sort the props? I think this line goes first.

File: src/core/components/Root/index.js (hunk 5 above)
Comment (willdurand) [reply src/core/components/Root/index.js]: weird.

Comment (willdurand) [reply src/core/components/Root/index.js]: Maybe you could put everything on a single line?

```js
{_config.get('enableStrictMode') ? <React.StrictMode>{children}</React.StrictMode> : children}
```

File: src/core/components/Root/index.js (hunk 4 above)
Comment (SeanPrashad) [reply src/core/components/Root/index.js]: It looks like Flow is blowing up now:

image


Should I go ahead and add this into `stories/setup/Provider.js`?

File: stories/setup/Provider.js [hunk 8]
```diff
@@ -14,7 +15,7 @@ const history = addQueryParamsToHistory({
 
 export default function Provider({ story }: Object) {
   return (
-    <Root store={store} history={history} i18n={fakeI18n}>
+    <Root store={store} history={history} i18n={fakeI18n} _config={config}>
```
Comment (willdurand) [inline stories/setup/Provider.js]: That's not needed. If Flow complains, then make the prop optional in `Root` using `_config?: typeof config`.

File: src/core/components/Root/index.js [hunk 9]
```diff
@@ -3,27 +3,40 @@ import * as React from 'react';
...
 |};
 
-const Root = ({ children, history, i18n, store, cookies = null }: Props) => (
+const Root = ({
+  _config = config,
+  children,
+  history,
+  i18n,
+  store,
+  cookies = null,
```
Comment (SeanPrashad) [inline src/core/components/Root/index.js]: I didn't want to touch this without asking, but should `cookies` go after `children` if doing ordering?

**Pull request merged**